### PR TITLE
PR: Optimise Spyder

### PR DESCRIPTION
**Pull request by impact27**
## Description of Changes
Small changes that improve the response time of spyder to keystrokes.

* [ ] Wrote at least one-line docstrings (for any new functions)
* [ ] Added unit test(s) covering the changes (if testable)

* [ ] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

### Issue(s) Resolved

Fixes #

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct:

### CODE REVIEW

**Comment by CAM-Gerlach** [comment]
Thanks, @impact27 ! Would be cool to see some benchmarks, if there's anything consistently measurable you can find.

**Comment by impact27** [comment]
Do you know of a good way of including benchmarks? It could be interesting to keep track of a few things in the CI maybe?

**Comment by impact27** [comment]
In the internal console:
This PR:
```
# Running in codeeditor.py
>>> timeit.timeit("spy.window.editor.zoom(1); spy.window.editor.zoom(-1)", number=3, globals={"spy":spy})
4.640133099999957
>>> timeit.timeit("PyQt5.QtTest.QTest.keyClick(spy.window.editor.get_focus_widget(), 'w')", number=100, globals={"spy":spy, "PyQt5": PyQt5})
0.5205837999999403
# Running in empty file
>>> timeit.timeit("spy.window.editor.run_cell()", number=100, globals={"spy":spy})
0.024335500000006505
```
5.x
```
# Running in codeeditor.py
>>> timeit.timeit("spy.window.editor.zoom(1); spy.window.editor.zoom(-1)", number=3, globals={"spy":spy})
7.019444599999986
>>> timeit.timeit("PyQt5.QtTest.QTest.keyClick(spy.window.editor.get_focus_widget(), 'w')", number=100, globals={"spy":spy, "PyQt5": PyQt5})
2.3932132999999993
# Running in empty file
>>> timeit.timeit("spy.window.editor.run_cell()", number=100, globals={"spy":spy})
1.7612139999999954
```

So 1.5x speed-up for zooming, 4.5x speed up for writing text, 72x speed up for running cells (only the overhead)

**Comment by impact27** [comment]
https://github.com/spyder-ide/spyder/pull/15372 is already useful :)

**Comment by CAM-Gerlach** [comment]
> Do you know of a good way of including benchmarks? It could be interesting to keep track of a few things in the CI maybe?

[Airspeed Velocity](https://asv.readthedocs.io/en/stable/); I think this is what CPython, PyPy and some other scientific packages use

> So 1.5x speed-up for zooming, 4.5x speed up for writing text, 72x speed up for running cells (only the overhead)

Nice! Glad to see; its not always easy since some PRs in the past that we thought would improve performance had no effect, or sometimes even caused it to regress, heh

**Comment by CAM-Gerlach** [comment]
As a quick, informal but practical/functional user test, I also often measure the max typing speed holding down one key (chars/s over 100 chars; which, at least in past versions, could vary by several times and be heavily constrained by Spyder performance rather than key repeat rate), the rate of moving lines up or down (lines per second over 1000 lines), and the undo/redo rate, plus a few others. These metrics are key in regulating how "snappy" and responsive Spyder feels to an average user, and historically have been many times lower than other editors (e.g. Notepad++).

**Comment by impact27** [comment]
Fast:
```
>>> timeit.timeit("PyQt5.QtTest.QTest.keyClick(spy.window.editor.get_focus_widget(), PyQt5.QtCore.Qt.Key_Enter)", number=100, globals={"spy":spy, "PyQt5": PyQt5})
3.0736162999999124
>>> timeit.timeit("spy.window.editor.get_current_editor().undo(); spy.window.editor.get_current_editor().redo()", number=100, globals={"spy":spy})
5.136109799999986
```

5.x
```
>>> timeit.timeit("PyQt5.QtTest.QTest.keyClick(spy.window.editor.get_focus_widget(), PyQt5.QtCore.Qt.Key_Enter)", number=100, globals={"spy":spy, "PyQt5": PyQt5})
4.196703599999864
>>> timeit.timeit("spy.window.editor.get_current_editor().undo(); spy.window.editor.get_current_editor().redo()", number=100, globals={"spy":spy})
6.325618500000019
```

1.4x faster for returns, 1.25 for undo/redo

**Comment by impact27** [comment]
With caching of bookmarks and breakpoints:
```
>>> timeit.timeit("spy.window.editor.get_current_editor().undo(); spy.window.editor.get_current_editor().redo()", number=100, globals={"spy":spy})
3.349273999999994
```
2x faster undo/redo

**Comment by CAM-Gerlach** [comment]
Nice! Fantastic, @impact27 .

**Comment by ccordoba12** [comment]
Hey @impact27, thanks for all your work on this! It's really impressive!!

For now I only have one suggestion for you: please remove from your last two commits (i.e. cc27158 and 
4231213) the changes related to `document_did_change`. The thing is right now I'm also working on using a timer to trigger `didChange` requests, but my approach is different from yours and I think it's the right one. I mean, I also tried your approach and unfortunately it can make that the file contents are not synced correctly with the contents saved in the PyLSP (which could be the cause of our tests failures).

Instead of calling `document_did_change` to update to the document contents manually, we can connect the timer to the `textChanged` signal of `QPlainTextEdit`, make the timer call `document_did_change` and remove all calls to that method. That solved the unsynced problems for me.

**Comment by impact27** [comment]
> Instead of calling `document_did_change` to update to the document contents manually, we can connect the timer to the `textChanged` signal of `QPlainTextEdit`, make the timer call `document_did_change` and remove all calls to that method. That solved the unsynced problems for me.

So you would still use a timer? I think the problem is the ordering of the calls to LSP. Doing them syncronously with the KeyPressEvent introduces a lot of percieved lag. I tried to only put document_did_change on a timer but I think there is an issue with the ordering of the calls.  So I see two problems:
 - The calls to LSP should not be syncronous with the KeyPressEvent 
 - The ordering of the calls should be preserved

**Comment by impact27** [comment]
@ccordoba12 I tried something here: https://github.com/spyder-ide/spyder/pull/16911

**Comment by ccordoba12** [comment]
> I tried something here: #16911

Thanks for your help with that @impact27! Let's continue the discussion about those improvements on that PR.

File: spyder/widgets/mixins.py [hunk 1]
```diff
@@ -38,6 +38,25 @@
...
+    "\x0b",  # Line Tabulation
+    "\f",  # Form Feed
+    "\x0c",   # Form Feed
+    "\x1c",   # File Separator
+    "\x1d",   # Group Separator
+    "\x1e",   # Record Separator
+    "\x85",   # Next Line (C1 Control Code)
+    "\u2028",   # Line Separator
+    "\u2029",   # Paragraph Separator
+    ]
```
Comment (ccordoba12) [inline spyder/widgets/mixins.py:57]:
```suggestion
]
```

Minor style issue.

**Review by ccordoba12 — COMMENTED** [review]
Thanks a lot for all these improvements @impact27! Apart from the minor comments I left below, they look really good to me.

File: spyder/plugins/editor/widgets/base.py [hunk 2]
```diff
@@ -478,12 +495,11 @@ def toPlainText(self):
         # corruptions when saving files with certain combinations
         # of unicode chars on them (like the one attached on
         # spyder-ide/spyder#1546).
+        text = super(TextEditBaseWidget, self).toPlainText()
```
[the remark this answers is not in the document]
Comment (impact27) [reply spyder/plugins/editor/widgets/base.py]: I couldn't download the file in the liked issue. I reverted this change

**Review by steff456 — CHANGES_REQUESTED** [review]
I tested this PR manually and I also left a couple of comments but this work was necessary! 

A thing that I noticed while playing around in this branch is that the highlight is taking some time to paint, I'm not sure if this has something to do with the Timer you added. I tried it in both a small and large file. 

find2

Also, I'm not sure why once I use the find option and delete the text in the find, the text is still highlighted until I focus the editor and type something,

find

**Review by CAM-Gerlach — COMMENTED** [review]
Great job, @mrclary !

I did observe the same behavior that @steff456 did, on Windows 8.1 x64/Py39 with clean settings, but it occurs on both `5.x` and this PR, so it doesn't seem like it was introduced here.

On 5.x, other than the first match nearest my cursor, which is highlighted instantly, there is around a 1.2-1.5 s delay between pasting a moderately common pattern (a dozen instances in a 200 line file) into the search box. On this PR, it was essentially identical, 1.3-1.6 s, so no apparent change here.

On both `5.x` and this PR, entirely clearing the find box (either by deleting the last character, or having all of them selected and pressing delete) does not clear the highlights, as @steff456 mentioned; even when clicking around in the editor; only modifying the text triggers a repaint and clears them.

Also, painting matches when selecting a particular name in the text and other instances being highlighted is delayed by 1.85 s +/- 0.2 s, which is consistent across both this PR and `5.x`.

I also noticed moderately odd behavior with the highlights being set to the foreground and background colors; when clicking off of the search box they are set to the bg selection color, but they don't get set back to the fg selection color if I click back in the search box, unless I change the text. However, yet again, this was the same between this PR and `5.x`

Here were my benchmark results, via real-world functional user testing, on line 1000 of `mainwindow.py` as always:

Key repeat rate:
5.x: 11 char/s
PR: 21 char/s

Line move rate
5.x: 6.5 lines/s
PR: 11 lines/s

Undo rate
5.x: 6.5 lines/s
PR: 10 lines/s

So, a substantial overall improvement, from 1.5x to 2x in the limited number of basic interaction tests I did.

When moving lines and undoing, I noticed several hundred ms pauses every 1.5-2 s (at a similar frequently to the refresh delay above) during redraws, which was fairly consistent (or even a bit longer) on this PR vs 5.x, but they seemed less frequent (I didn't rigorously measure this) and the move speed between redraws was much faster, indicating less stuff was getting done with each line change.

File: spyder/plugins/editor/widgets/codeeditor.py [hunk 3]
```diff
@@ -569,6 +574,11 @@ def __init__(self, parent=None):
         self.is_undoing = False
         self.is_redoing = False
 
+        # Timer to Avoid too many calls to rehighlight.
+        self._rehighlight_timer = QTimer(self)
+        self._rehighlight_timer.setSingleShot(True)
+        self._rehighlight_timer.setInterval(150)
```
[the remark this answers is not in the document]
Comment (impact27) [reply spyder/plugins/editor/widgets/codeeditor.py:580]: So you need to wait long enough that two consecutive scrolling events are merged together but not long enough that it introduces a noticable delay. The human mind can detect a delay of about .1s so this seems about right

File: spyder/plugins/editor/widgets/editor.py [hunk 4]
```diff
@@ -1496,11 +1496,19 @@ def has_filename(self, filename):
...
-        for index, finfo in enumerate(self.data):
-            if fixpath(filename) == fixpath(finfo.filename):
-                return index
-        return None
+        data_filenames = self.get_filenames()
+        try:
+            # Try finding without calling the slow realpath
+            return data_filenames.index(filename)
+        except ValueError:
+            def fixpath(path):
```
[the remark this answers is not in the document]
Comment (impact27) [reply spyder/plugins/editor/widgets/editor.py:1504]: I think there should be a fix where all the filenames are validated before being used in the editor. But this is a bit out of scope for this PR.

[the remark this answers is not in the document]
Comment (steff456) [reply spyder/plugins/editor/widgets/editor.py:1504]: I understand that, but I meant that I will prefer to not have nested functions because usually they are hard to follow and in this particular case you don't absolutely need the nested function there.

**Review by dalthviz — COMMENTED** [review]
Checking this I encounter with some sort of bug with the indentation guides when adding some lines at the top of a file (although not related with this since I was able to reproduce it with 5.x):

indentguides

But overall seems like the Editor response is better (when checking the zoom behavior now it can handle better quick zoom in/zoom out changes :+1:)

File: spyder/plugins/editor/widgets/editor.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (impact27) [reply spyder/plugins/editor/widgets/editor.py]: I'll fix that

**Comment by ccordoba12** [comment]
@impact27, I think you still need to address the two regressions @steff456 mentioned above:

> A thing that I noticed while playing around in this branch is that the highlight is taking some time to paint

Perhaps for this we could be solved by shortening the timeout of `_rehighlight_timer`. Or we could even leave that for another PR to investigate further how to improve it.

> Also, I'm not sure why once I use the find option and delete the text in the find, the text is still highlighted until I focus the editor and type something,

This is a serious UI/UX regression. Perhaps it's related to the introduction of `_rehighlight_timer` too (not really sure though).

**Comment by CAM-Gerlach** [comment]
@ccordoba12 , as I stated and I believe @dalthviz confirmed in our reviews, both of those do not appear to be regressions with this PR, as they were identically reproducible on `5.x`, with the same behavior and lag time observed down to the measuring accuracy (0.2-ish s). I would suggest opening new issues about those, including the information in @steff456 's and my reports.

**Comment by ccordoba12** [comment]
> Actually @ccordoba12 , as I stated and I believe @dalthviz confirmed in our reviews, both of those do not appear to be regressions with this PR, as they were identically reproducible on 5.x

Oh, so sorry! I'll try to read your comments more carefully for next time.

Then I'll wait for another review from @steff456 and @dalthviz. If they approve this one, I'll merge it.

**Comment by CAM-Gerlach** [comment]
No worries @ccordoba12 ^_^ , I realize I could have made it more clear. Do you want me to open new issues about the UI issues @steff456 , @dalthviz and I found? NB, I can repro them on 5.1.5 as well. I think the first thing is not really an issue though, since IIRC based on my feedback you set the timer like that (1s-ish timed from keyboard idle, I believe) a while ago to prevent major searching/painting lag when typing common search queries in longer documents. It personally doesn't really bother me too much, though a bit shorter might be better without affecting responsiveness too much.

**Comment by ccordoba12** [comment]
> Do you want me to open new issues about the UI issues @steff456 , @dalthviz and I found?

Yes, please do. Thanks for your help with that!

**Comment by CAM-Gerlach** [comment]
Thanks! Opened as #16948 , I imagine @steff456 or @impact27 might be good people to look into it

**Review by steff456 — APPROVED** [review]
This looks ready to merge!

**Review by ccordoba12 — APPROVED** [review]
Thanks @impact27 for your work on this, and @steff456 and @dalthviz for your reviews. Great work here!

**Comment by CAM-Gerlach** [comment]
I see how it is @ccordoba12 :sob: 

In all seriousness, thanks @impact27 and awesome to see this merged!

**Comment by ccordoba12** [comment]
Haha, sorry, I forgot about your review. Thanks for your contribution too!
